### PR TITLE
Add support for QuantaMesh Common Rangeley CPU board

### PR DESCRIPTION
**Pull request by audihsu-qci**
===============================================================================
  QuantaMesh Common Rangeley CPU board
===============================================================================

  QuantaMesh switch which supports modularized CPU board design, have a
non-programmable H/W component for mother board detection. For motherboard
detection, please refer to ``busybox/miscutils/mb_detect.c``

  Supported Platforms
=======================
 - T5032-LY6  (cpu module)
   32x40GbE QSFP+ ports

 - T3048-LY8  (cpu module)
   48x1/10GbE SFP+ ports,
   front 4x40GbE QSFP+ ports, and rear 2x40GbE QSFP+ ports

 - T3048-LY5A (cpu module)
   48x1/10GbE SFP+ ports and 12x40GbE QSFP+ ports

 - T3048-LY9  (cpu module)
   48x100/1000/10GBASE-T,
   front 4x40GbE QSFP+ ports, and rear 2x40GbE QSFP+ ports

 - T3048-LY9A (cpu module)
   48x100/1000/10GBASE-T and 6x40GbE QSFP+ ports

 - T5032-LY6F (cpu module)
   32x40GbE QSFP+ ports

 - For more detail product information, please refer to
   http://www.quantaqct.com/

  Specification
===============
 - Modularized CPU board design
 - 1 RJ-45 out-of-band management port (10/100/1000M)
 - 1 RJ-45 console port
 - 1 4GB DDR3 SDRAM
 - 1 mSATA 32GB
 - 1 USB 2.0 port

For further information, please take a look at machine/quanta/quanta_common_rangeley/INSTALL

### CODE REVIEW

File: machine/quanta/quanta_common_rangeley/busybox/patches/onie-sys-eeprom-platform.patch [hunk 1]
```diff
@@ -0,0 +1,722 @@
...
+
+diff --git a/include/applets.src.h b/include/applets.src.h
+index 1c8e0cb..a600632 100644
+--- a/include/applets.src.h
++++ b/include/applets.src.h
+@@ -417,6 +417,7 @@ IF_YES(APPLET_NOFORK(yes, yes, BB_DIR_USR_BIN, BB_SUID_DROP, yes))
+ IF_GUNZIP(APPLET_ODDNAME(zcat, gunzip, BB_DIR_BIN, BB_SUID_DROP, zcat))
+ IF_ZCIP(APPLET(zcip, BB_DIR_SBIN, BB_SUID_DROP))
+ IF_SYS_EEPROM(APPLET_ODDNAME(onie-syseeprom, onie_syseeprom, BB_DIR_USR_BIN, BB_SUID_DROP, onie_syseeprom))
++APPLET_ODDNAME(mb_detect, mbdetect, BB_DIR_USR_BIN, BB_SUID_DROP, mbdetect)
```
Comment (cbrune) [inline machine/quanta/quanta_common_rangeley/busybox/patches/onie-sys-eeprom-platform.patch]: Does this need a `IF_MBDETECT()` wrapper?  Like the line above related to `CONFIG_SYS_EEPROM`.

File: machine/quanta/quanta_common_rangeley/busybox/patches/onie-sys-eeprom-platform.patch [hunk 2]
```diff
@@ -0,0 +1,722 @@
...
++		}
++
++		return 0;
++	}
++
++	ep = NULL;
++	return -1;
++}
++
++#if 0
```
Comment (cbrune) [inline machine/quanta/quanta_common_rangeley/busybox/patches/onie-sys-eeprom-platform.patch]: remove this dead code

File: machine/quanta/quanta_common_rangeley/installer.conf [hunk 3]
```diff
@@ -0,0 +1,15 @@
+# Quanta Common Rangeley CPU board
+
+description="Quanta, Common Rangeley"
+
+# Default ONIE block device
+install_device_platform()
+{
+	echo /dev/sda
```
Comment (cbrune) [inline machine/quanta/quanta_common_rangeley/installer.conf]: Will this always be true?  Even when a USB memory stick is inserted?

File: machine/quanta/quanta_common_rangeley/kernel/config [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+CONFIG_VIRTIO_RING=y
+CONFIG_VIRTIO_PCI=y
+CONFIG_VIRTIO_BALLOON=y
+CONFIG_PARAVIRT_GUEST=y
+CONFIG_PARAVIRT=y
+CONFIG_PARAVIRT_SPINLOCKS=y
+CONFIG_PARAVIRT_CLOCK=y
+CONFIG_KVM_CLOCK=y
+CONFIG_KVM_GUEST=y
+CONFIG_HAVE_KVM=y
```
Comment (cbrune) [inline machine/quanta/quanta_common_rangeley/kernel/config]: Why do you enable virtualization support?

File: machine/quanta/quanta_common_rangeley/kernel/config [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+# CONFIG_DMADEVICES_DEBUG is not set
+# CONFIG_INTEL_MID_DMAC is not set
+CONFIG_INTEL_IOATDMA=y
+# CONFIG_TIMB_DMA is not set
+# CONFIG_PCH_DMA is not set
+CONFIG_DMA_ENGINE=y
+CONFIG_NET_DMA=y
+# CONFIG_ASYNC_TX_DMA is not set
+# CONFIG_DMATEST is not set
+CONFIG_DCA=y
```
Comment (cbrune) [inline machine/quanta/quanta_common_rangeley/kernel/config]: Questions on the kernel config:

1. why enable the DMA support?
2. why is DCA needed?
3. why is INTEL_IOATDMA needed?

File: machine/quanta/quanta_common_rangeley/kernel/driver-support-broadcom-phy-bcm5461s.patch [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+@@ -2120,6 +2120,36 @@ s32 igb_phy_init_script_igp3(struct e1000_hw *hw)
+ }
+ 
+ /**
++ *  igb_phy_init_script_5461s - Inits the BCM5461S PHY
++ *  @hw: pointer to the HW structure
++ *
++ *  Initializes a Broadcom Gigabit PHY.
++ **/
++s32 igb_phy_init_script_5461s(struct e1000_hw *hw)
```
Comment (cbrune) [inline machine/quanta/quanta_common_rangeley/kernel/driver-support-broadcom-phy-bcm5461s.patch]: Why is this function needed?

The e1000 should be configured by its EEPROM.  That is the usual method for setting board specific configuration parameters.

File: machine/quanta/quanta_common_rangeley/machine.make [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+# Vendor ID -- IANA Private Enterprise Number:
+# http://www.iana.org/assignments/enterprise-numbers
+# DELL IANA number
+VENDOR_ID = 7244
+
+I2CTOOLS_ENABLE = yes
+I2CTOOLS_SYSEEPROM = no
+
+# This platform requires the PXE_EFI64 installer
+PXE_EFI64_ENABLE = no
```
Comment (cbrune) [inline machine/quanta/quanta_common_rangeley/machine.make]: The setting and the comment above seem to conflict -- which is it?

File: machine/quanta/quanta_common_rangeley/machine.make [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+CONSOLE_SPEED = 115200
+CONSOLE_DEV = 1
+CONSOLE_FLAG = 0
+
+EXTRA_CMDLINE_LINUX = i2c-ismt.bus_speed=100
```
Comment (cbrune) [inline machine/quanta/quanta_common_rangeley/machine.make]: Why do you need this setting?

File: machine/quanta/quanta_common_rangeley/rootconf/sysroot-init/rtc-sync.sh [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+	mkdir /sysfs
+	mount -t sysfs sysfs /sysfs
+	/bin/echo ds1339 0x68 > /sysfs/devices/pci0000:00/0000:00:1f.3/i2c-0/new_device
+	umount /sysfs
+	rm -rf /sysfs
+	/bin/mknod /dev/rtc1 c 254 1
+	/bin/ln -sf /dev/rtc1 /dev/rtc
+	/sbin/hwclock --hctosys
+	log_info_msg "Initializing RTC DS1339 Done ..."
+}
```
Comment (cbrune) [inline machine/quanta/quanta_common_rangeley/rootconf/sysroot-init/rtc-sync.sh]: Why is this necessary?  x86 systems use the CMOS RTC.  The kernel drivers for that are already in place and should just work.

File: machine/quanta/quanta_common_rangeley/busybox/patches/onie-sys-eeprom-platform.patch (hunk 1 above)
Comment (audihsu-qci) [inline machine/quanta/quanta_common_rangeley/busybox/patches/onie-sys-eeprom-platform.patch]: CONFIG_MBDETECT=y will not applied to busybox .config because ``apply-config-patch`` script will filter this out.
I finally chose to remove the IF_XXXX prefix and make sure MBDETECT works.

File: machine/quanta/quanta_common_rangeley/kernel/driver-support-broadcom-phy-bcm5461s.patch (hunk 6 above)
Comment (audihsu-qci) [inline machine/quanta/quanta_common_rangeley/kernel/driver-support-broadcom-phy-bcm5461s.patch]: BCM5461S LED polarity is not correct by default. This function sets the LED speed/activity to which Quanta defines 1000M -> Orange, 100M -> Green, 10M -> Off. Any better suggestions to apply this?

File: machine/quanta/quanta_common_rangeley/rootconf/sysroot-init/rtc-sync.sh (hunk 9 above)
Comment (audihsu-qci) [inline machine/quanta/quanta_common_rangeley/rootconf/sysroot-init/rtc-sync.sh]: Quanta chose an alternate RTC, which is Maxim DS1339. The same as we did in PPC CPU board.
The default CMOS RTC does not work well.
Therefore, before file system is mounted, add DS1339 as default RTC.

**Comment by audihsu-qci** [comment]
Hi Curt,
    Thanks for your review and suggestions.

File: machine/quanta/quanta_common_rangeley/kernel/driver-support-broadcom-phy-bcm5461s.patch (hunk 6 above)
Comment (cbrune) [inline machine/quanta/quanta_common_rangeley/kernel/driver-support-broadcom-phy-bcm5461s.patch]: The better thing to do is update your manufacturing process to program the e1000 configuration eeprom with the necessary board specific register configurations.  Doing it in software is a long term losing battle as the code will never be accepted by the upstream Linux kernel developers.

File: machine/quanta/quanta_common_rangeley/rootconf/sysroot-init/rtc-sync.sh (hunk 9 above)
Comment (cbrune) [inline machine/quanta/quanta_common_rangeley/rootconf/sysroot-init/rtc-sync.sh]: That is a really unfortunate hardware design for x86.

This puts additional requirements on the kernel as the RTC *should* be properly configured into the kernel and the RTC programmed by the *kernel* at boot time.  Programming the RTC from user space after the system as booted is a pretty big hack.

For example the initial dmesg log messages will have the wrong date/time before the RTC is fixed up from userspace.

Please encourage your hardware team to do it the standard x86 way in the future.

File: machine/quanta/quanta_common_rangeley/machine.make (hunk 8 above)
Comment (cbrune) [inline machine/quanta/quanta_common_rangeley/machine.make]: Why do you need `i2c-ismt.bus_speed=100` ?

File: machine/quanta/quanta_common_rangeley/kernel/driver-support-broadcom-phy-bcm5461s.patch (hunk 6 above)
Comment (QuantaSwitchONIE) [inline machine/quanta/quanta_common_rangeley/kernel/driver-support-broadcom-phy-bcm5461s.patch]: Hi Curt,

I don’t think a mii cmd init stript can be done in Rangeley eeprom, as I know, it only config for sgmii/serdes/kx/pci/phyid
PHY default behavior  can be set on BCM5461s eeprom, but unfortunately we don’t have a 5461 eeprom on it.

Usually,  bootloader will also set the LED behavior, but during KERNEL Ethernet driver init, it will reset phy, so the LED configuraton will become default, that’s why we need to run a init script in driver after phy init.

Look at the original igb driver code style, there is a igb_phy_init_script_igp3,
it is design for support different phy init script usage.

File: machine/quanta/quanta_common_rangeley/machine.make (hunk 8 above)
Comment (audihsu-qci) [inline machine/quanta/quanta_common_rangeley/machine.make]: This is a porting mistake, I removed it.
I shall close this session and have another pull request for all these changes.
